odxtools/cli/browse.py

```diff
@@ -192,56 +207,58 @@ def encode_message_from_string_values(
     if parameter_values is None:
         parameter_values = {}
     parameter_values = parameter_values.copy()
-    param_dict = sub_service.parameter_dict()
 
-    # Check if all needed parameters are given
+    # Check if all needed parameters have been specified
     missing_parameter_names = []
-    for param_sn, param in param_dict.items():
-        if isinstance(param, dict):
-            # param_value refers to a structure (represented as dict of params)
-            for simple_param_sn, simple_param in param.items():
-                structured_value = parameter_values.get(param_sn)
-                if not isinstance(simple_param, Parameter):
-                    continue
-                if simple_param.is_required and (not isinstance(structured_value, dict) or
-                                                 structured_value.get(simple_param_sn) is None):
-                    missing_parameter_names.append(f"{param_sn} :: {simple_param_sn}")
+    for param in sub_service.parameters:
+        if (inner_params := getattr(dop := getattr(param, "dop", None), "parameters",
+                                    None)) is not None:
+            inner_param_values = parameter_values.get(param.short_name, {})
+            if not isinstance(inner_param_values, dict):
+                print(f"Value for composite parameter {param.short_name} must be "
+                      f"a dictionary, got {type(inner_param_values).__name__}")
+                continue
+            for inner_param in inner_params:
+                if inner_param.is_required and inner_param.short_name not in inner_param_values:
+                    missing_parameter_names.append(f"{param.short_name}.{inner_param.short_name}")
         else:
-            if param.is_required and parameter_values.get(param_sn) is None:
-                missing_parameter_names.append(param_sn)
+            if param.is_required and parameter_values.get(param.short_name) is None:
+                missing_parameter_names.append(param.short_name)
 
     if len(missing_parameter_names) > 0:
-        print("The following parameters are required but missing!")
-        print(" - " + "\n - ".join(missing_parameter_names))
+        print("The following parameters are required but missing:")
+        print(" - " + "\n - ".join(sorted(missing_parameter_names)))
         return
 
     # Request values for parameters
     for parameter_sn, parameter_value in parameter_values.items():
-        parameter = param_dict.get(parameter_sn)
+        parameter = resolve_snref(parameter_sn, sub_service.parameters, Parameter)
         if parameter is None:
             print(f"I don't know the parameter {parameter_sn}")
             continue
 
         if isinstance(parameter_value, dict):
             # parameter_value refers to a structure (represented as dict of params)
+            dop = getattr(parameter, "dop", None)
+            inner_params = getattr(dop, "parameters", None)
```

Comment:
not using the interface you created?

Reply:
not every parameter exhibits a DOP (e.g. `ReservedParameter`) and not every DOP has parameters (only `ComplexDop`s like structures and environment datas)...

Reply:
is `isintance(dop, CompositeCodec)` more performant than `getattr`? (I assume that it is significantly slower)

Reply:
when it comes to performance, the only way to know for sure is to measure

Reply:
if I did not make a mistake, `getattr` is faster by 5 orders of magnitude:

```
> cat bla.py
import odxtools
import timeit

from typing import Optional, List
from odxtools.codec import CompositeCodec
from odxtools.odxtypes import ParameterValue
from odxtools.decodestate import DecodeState
from odxtools.encodestate import EncodeState
from odxtools.parameters.parameter import Parameter
from odxtools.codec import Codec
from odxtools.codec import CompositeCodec

class NotACodec:
    def foo(self):
        pass

class ABasicCodec(NotACodec):
    @property
    def short_name(self) -> str:
        return "foo"

    def encode_into_pdu(self, physical_value: Optional[ParameterValue],
                        encode_state: EncodeState) -> None:
        pass

    def decode_from_pdu(self, decode_state: DecodeState) -> ParameterValue:
        ...

    def get_static_bit_length(self) -> Optional[int]:
        ...

class ACompositeCodec(ABasicCodec):
    @property
    def parameters(self) -> List[Parameter]:
        return []

    @property
    def required_parameters(self) -> List[Parameter]:
        return []

    @property
    def free_parameters(self) -> List[Parameter]:
        return []


a = NotACodec()
b = ABasicCodec()
c = ACompositeCodec()

n = 100_000

def foo():
    for i in range(n):
        isinstance(c, CompositeCodec)

def bar():
    for i in range(n):
        getattr(c, 'parameters')

t1 = timeit.timeit("foo()", number=1, globals=globals())
print(f"isinstance: {t1}")
t2 = timeit.timeit("bar()", number=1, globals=globals())
print(f"getattr: {t2}")
> python3 bla.py 
isinstance: 1.728830883000228
getattr: 0.008215672000005725
```